### PR TITLE
Added a script and a Dockerfile to run with beaker

### PR DESCRIPTION
**Pull request by matt-gardner**
This is by no means perfect, but it's what I used to get AllenNLP working with beaker.  Feedback welcome.

A couple of questions:

1. Seems like it might be useful to just create a directory for dockerfiles, so we can provide several for different use cases.  We currently have two, this PR makes it three, and it's not hard to imagine this number growing in the future.  Maybe they should be `docker/cpu_dev`, `docker/gpu_dev`, `docker/gpu_train`, etc.?  Any other ideas?

2. The beaker script is a bit of a mess.  I don't really want to make it nicer in bash; I'd probably convert it to python if we want to make it nicer.  Also not sure where it should live.  I don't have a problem with it being in the `scripts/` directory, as long as it's marked as being AI2-specific (maybe `scripts/ai2-internal/`?), but I can understand not wanting to have it there.  I don't have a good alternative, though.

### CODE REVIEW

**Comment by schmmd** [comment]
Ideally you could just `allenai-gpu` and pass in the command used to train.  My intention was to have only two Dockerfiles (one for cpu one for gpu) and to pass whatever command is necessary.

I'd love to start moving internal scripts into a particular folder.  It will make it easier to assess what AI2-specific items we have. `/scripts-internal` or `/scripts/ai2-internal` both seem like fine candidates.

File: scripts/run_with_beaker.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+fi
+
+
+set -e
+
+# Get temporary ecr login. For this command to work, you need the python awscli
+# package with a version more recent than 1.11.91.
+eval $(aws --region=us-west-2 ecr get-login --no-include-email)
+
+docker build -f Dockerfile.gpu.train -t $IMAGE . --build-arg PARAM_FILE=$PARAM_FILE
```
Comment (schmmd) [inline scripts/run_with_beaker.sh]: @splittingfield Matt creates a dockerfile here with the `CMD` and `model_params.json` he needs to train his model.  Is it possible to somehow instead use `Dockerfile.gpu` and add a command and `model_params.json`?  Or should we base `Dockerfile.gpu.train` on `Dockerfile.gpu`?

Comment (schmmd) [inline scripts/run_with_beaker.sh]: I would think it's possible to pass in the data `model_params.json` somehow, and also specify the command that Beaker runs in the Docker image.

File: scripts/run_with_beaker.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+GPU_ARG="--gpu-count=1"
+DETACH_ARG="--detach"  # or ""
+
+ECR_REPOSITORY=896129387501.dkr.ecr.us-west-2.amazonaws.com
+
+PARAM_FILE=$1
+EXPERIMENT_NAME=$2
+
+COMMIT=$(git rev-parse HEAD)
+IMAGE=$ECR_REPOSITORY/allennlp/allennlp-gpu:$COMMIT
```
Comment (joelgrus) [inline scripts/run_with_beaker.sh]: if two people check out HEAD and make changes without committing, their images will step on each other? do we want to add more protection for this?

File: scripts/run_with_beaker.sh (hunk 1 above)
Comment (joelgrus) [inline scripts/run_with_beaker.sh]: it might be possible to do that with mounts? feels kind of hacky, though

Comment (matt-gardner) [inline scripts/run_with_beaker.sh]: I don't think you could do it with mounts, because the file doesn't exist on the kube machine where this is run.

File: scripts/run_with_beaker.sh (hunk 2 above)
Comment (schmmd) [inline scripts/run_with_beaker.sh]: You could add a check to make sure all changes are committed.  See my `run_on_kube.sh` for the code.

**Comment by schmmd** [comment]
Most projects have a single Dockerfile at the root of the project.  AllenNLP is a bit unusual in that we have two--but I think that it makes sense as we have two separate roots for GPU and CPU support.  We could in theory build the GPU image from the CPU image, but that would take some serious work to re-implement GPU support and it might take a long time to build the GPU image.

I think it is very unfortunate if we need to add a third Dockerfile, which is essentially a copy of `Dockerfile.gpu`.  It makes it even harder to change our Dockerfiles, because now most changes need to be made in three places.

If there is no mounting/downloading solution then I would suggest building `Dockerfile.gpu.train` from `Dockerfile.gpu`.  Here is some pseudocode about how your script would change.

```
docker build -f Dockerfile.gpu -t allennlp-gpu
docker build -f Dockerfile.gpu.train -t allennlp-gpu-train # this will extend the previous step
```

However, I know we're in a crunch.  If you need to merge forward, I could take a look at how to organize this.

File: scripts/run_with_beaker.sh (hunk 2 above)
Comment (schmmd) [inline scripts/run_with_beaker.sh]: I also worry about the potential confusion of an image labeled with a git commit, but if there are local changes the image and the commit won't be aligned.

File: scripts/run_with_beaker.sh (hunk 1 above)
Comment (schmmd) [inline scripts/run_with_beaker.sh]: Does it make any sense to pass it as a source?

Comment (matt-gardner) [inline scripts/run_with_beaker.sh]: That means you have to create a beaker dataset for it, which doesn't seem to make much sense to me.

**Comment by schmmd** [comment]
You would probably need to generate the Dockerfile.gpu.train script from the bash script, since it will be built from whatever the preceeding image is named.

File: scripts/run_with_beaker.sh (hunk 2 above)
Comment (matt-gardner) [inline scripts/run_with_beaker.sh]: Yeah, I didn't copy that part of the code because modifying BiDAF config files makes the repo dirty, and I didn't want to have to make a commit just to run the experiment, especially because the docker image associated with an experiment _should_ be recoverable, and I can just inspect the code if I really want to know what was run.

As I said, this script is still a bit of a mess.  But I'm not sure I want to _require_ a clean repo to do this.  Maybe we can append a random number / string / hash to the commit id?

Comment (schmmd) [inline scripts/run_with_beaker.sh]: Adding a random number seems like a reasonable solution.  Kubernetes might require that it's all lowercase alphanumeric (like a git sha)--I had some problems earlier with appending random hashes.

**Comment by splittingfield** [comment]
General comment: The docker best practice here is to specify all config as environment variables.  Then Beaker can pick up these parameters and display them appropriately in the UI.

For situations where there is a required config file that could be big, the docker best practice (https://groups.google.com/forum/#!topic/docker-user/FyCWLC38Ueg) is to use a wrapper script as your entry point that 

1) takes the environment variables that need runtime modification 
2) generates the required config file
3) launches the appropriate task.  

For reproducibility you could write the generated config file to the output path to save.

@ckarenz may wish to comment on this as well, as he is the domain expert in docker. He also may have some ideas for a feature request that could somehow help streamline this process.

I believe Bhavana has a script that takes a config file, creates a Beaker dataset as input and then automatically uses that as a source.  This works, but will lose the ability to do hyper parameter search, as we just can't yet know how to format and parse that to generate new parameters.

File: scripts/run_with_beaker.sh (hunk 1 above)
Comment (schmmd) [inline scripts/run_with_beaker.sh]: Marc is giving this some thought.  He linked to https://groups.google.com/forum/#!topic/docker-user/FyCWLC38Ueg and explained that most docker parameterization is done with environment variables--although we do have a lot of configuration.

If we were running things locally with Docker, I would consider passing the parameters via stdin, but I don't think that's an option with beaker.

Comment (ckarenz) [inline scripts/run_with_beaker.sh]: You have a few options:
- Build in the config, rebuild the image each run. This is tremendously inefficient, but the most flexible.
- Mount (pull the config from a file passed in at runtime): In Beaker this would be a data set.
- Env var / command line args: Each option gets its own variable or arg.
- Config blob: Pass the json blob as an env var or command line arg. The AWS CLI does this.

I'd recommend a combination of things, depending on how much is truly configurable: Build in the most stable parts of config, and things that you change often would be environment variables or CLI args.

**Comment by ckarenz** [comment]
I've left an initial recommendation above, but it's tentative.  I'd like to better understand the config before making a firm statement.

File: scripts/run_with_beaker.sh (hunk 1 above)
Comment (matt-gardner) [inline scripts/run_with_beaker.sh]: This is an actual config: https://github.com/allenai/allennlp/blob/master/experiment_config/bidaf.json.  I'm not sure that any of it can be said to be "most stable".  Converting it into a set of environment variables seems infeasible (and would really mess with our code that reads parameters).  A json blob as an environment variable seems possible, maybe, though I'm not sure exactly what that would look like.

**Comment by schmmd** [comment]
@ckarenz let me grab some time with you tomorrow.

File: scripts/run_with_beaker.sh (hunk 1 above)
Comment (schmmd) [inline scripts/run_with_beaker.sh]: You already bundle `allennlp/` into the docker image.  We could either commit the experiment json or keep it in a `.gitigore`ed file and bundle it in the docker image.  That way it doesn't need to be passed as a separate file, although it would need to be an argument to the command.

This feel a bit hacky, but it would avoid the need for an additional Docker image.

**Comment by schmmd** [comment]
I discussed this with @ckarenz and we came up with three options.

1.  Use `Dockerfile.gpu` but put your experiments in a folder under `allennlp/` so they get packaged into the Dockerfile.  You can specify the command to run when you call Beaker.  This forces you to put your experiment files inside `allennlp/`, but has no other downsides.  You could create a `.gitignore`-ed folder if you want your experiments to be outside git.

2.  Mount your parameters file as a dataset.  We thought that felt weird earlier, but Colin found it completely normal/acceptable.

3.  Commit "default" JSON files in the repo, and then override them when running experiments with CLIs or ENVs.  This has the added advantage that beaker makes it easy to see configuration in the UI.

**Comment by splittingfield** [comment]
Following up here, do we need any more input to get this going?

**Comment by matt-gardner** [comment]
I've been putting out fires for the AllenNLP release, so I haven't touched this recently, even though it would have helped me put those fires out if it worked a little better - getting it to work better was going to take too long, so I just didn't use it. 

I'll take option 1 from @schmmd's list above, and copy the config file in the bash script from wherever it is to a common, ignored location in the current directory.

To make beaker actually useful, I still need to modify the training code to have an option to not use tqdm, because tqdm does not play nicely with beaker. I'm hoping that I just put out the last major fire, so I might be able to get to this tomorrow.

**Comment by splittingfield** [comment]
What is tqdm?  The progress meter?  We do have a feature request for
getting intermediate state. I believe Colin is putting together a plan for
logs right now.

**Comment by splittingfield** [comment]
ALso, just let us know how we can help.

**Comment by matt-gardner** [comment]
Also, on a related note, I'd say the two things that are my biggest requests for beaker right now:

1. A nice way to get a live dashboard on all currently running experiments, and/or to expose a tensorboard UI for a running / finished experiment. Michal said something about side car jobs that might make this possible.
2. A way other than through env / args to get "parameters" into the system. It could be that tags are sufficient here, if there's a way that I can have those tags automatically populated from a config file.

**Comment by splittingfield** [comment]
Ok.  Let's chat some this week.  I need to understand (2) a bit better and
for (1) does PyTorch support tensorboard?

**Comment by matt-gardner** [comment]
Yeah, tqdm is the progress bar, and doesn't output newlines, so logging doesn't show up in beaker until tqdm is finished (could be ~30 minutes or more).  Temporary fix for this is to just not use tqdm when using beaker, but I need to refactor code a bit for that, and add some parameters.

Yes, pytorch supports tensorboard.  Or, at least there are third-party libraries that let you use tensorboard with pytorch, and we're using one.

**Comment by splittingfield** [comment]
Meeting invite sent.  We can discuss off github comments... I have a
thought here that may make this straight forward for tensorboard at least.

**Comment by joelgrus** [comment]
@splittingfield can you add me to the invite too

**Comment by matt-gardner** [comment]
Ok, whoever's interested, this is ready for another look.  These are the changes since people last looked at it:

- Using a relatively new feature of docker, I was able to combine the two original dockerfiles into a single dockerfile, with an argument for what base to start from.  The default is to build for GPUs, and you can pass `--build-arg from=ubuntu:16.04` to get what used to be `Dockerfile.cpu`.  You must have docker >= 1.17 for this to work, I believe (at least it doesn't work on 1.13, and it looked like the next step up was 1.17).
- I removed the internal conda environment, because it's totally unnecessary - we just use the base environment as we're already inside a container.  This makes all of the `RUN` statements much simpler and easier to parse.
- I switched the `run_with_beaker.sh` script to pass the command to the dockerfile, instead of requiring the dockerfile to already have the command.  The script still needs some work to be more general, but it's at least a starting place if you want to run an experiment.
- I added an option to the training code to not use tqdm.

I successfully started a [beaker job](http://beaker.allenai.org/ex/ex_jn5mj85d2bg7) using this script, and it worked.  Because we're bypassing tqdm, the logs are actually useful for seeing what's going on.

File: Dockerfile [hunk 3]
```diff
@@ -32,24 +33,20 @@ RUN apt-get update --fix-missing && apt-get install -y \
...
 # We only copy what we need here so small changes to the repository does not trigger re-installation of the requirements.
 COPY requirements.txt .
 COPY requirements_test.txt .
 COPY scripts/install_requirements.sh scripts/install_requirements.sh
-RUN bash -c 'source activate allennlp && INSTALL_TEST_REQUIREMENTS="true" ./scripts/install_requirements.sh'
-RUN bash -c 'source activate allennlp && pip install --no-cache-dir -q http://download.pytorch.org/whl/cu80/torch-0.2.0.post1-cp35-cp35m-manylinux1_x86_64.whl'
+RUN INSTALL_TEST_REQUIREMENTS="true" ./scripts/install_requirements.sh
+RUN pip install --no-cache-dir -q http://download.pytorch.org/whl/cu80/torch-0.2.0.post1-cp35-cp35m-manylinux1_x86_64.whl
 
 COPY . .
```
Comment (matt-gardner) [inline Dockerfile]: We should copy all code necessary for tests here, then run tests, _then_ do `COPY . .`, so copying a new parameter file into `.beaker/model_params.json` doesn't trigger re-running the tests.

File: scripts/ai2-internal/run_with_beaker.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+# Get temporary ecr login. For this command to work, you need the python awscli
+# package with a version more recent than 1.11.91.
+eval $(aws --region=us-west-2 ecr get-login --no-include-email)
+
+mkdir -p .beaker/
+cp $PARAM_FILE .beaker/model_params.json
+docker build -t $IMAGE .
+docker push $IMAGE
+
+CMD="allennlp/run train .beaker/model_params.json"
```
Comment (matt-gardner) [inline scripts/ai2-internal/run_with_beaker.sh]: There's a decent argument for just making a dataset out of the model params file.  It would make it easier to access if you want to just inspect the parameter file, or use it as the base for a subsequent experiment.

On the other hand, we copy the parameter file to the serialization directory, so it's just as easily accessible in the result dataset.

**Comment by joelgrus** [comment]
did you see that tqdm has a `disable` option?

https://github.com/tqdm/tqdm/blob/master/tqdm/_tqdm.py#L617

so another possibility would be

```
self._no_tqdm = no_tqdm
```

then call `tqdm.tqdm` with `disable=self._no_tqdm`

and then still do the `logger.info` if tqdm is turned off.  might be cleaner?

**Comment by matt-gardner** [comment]
Thanks for the suggestion @joelgrus.  Fixed.

**Comment by joelgrus** [comment]
nice, that feels a lot cleaner
